triton-vm/src/table/hash_table.rs

```diff
@@ -450,36 +334,120 @@ impl ExtHashTable {
+        // round numbers evolve as
+        // 1 -> 2 -> 3 -> 4 -> 5 -> 6 -> 7 -> 8 -> 9, and
+        // 9 -> 1 or 9 -> 0, and
+        // 0 -> 0
+
+        // 1. round number belongs to {0, ..., 9}
+        // => consistency constraint
 
-        vec![
-            if_round_number_is_x_then_round_number_next_is_y(0, 0),
-            if_round_number_is_x_then_round_number_next_is_y(1, 2),
-            if_round_number_is_x_then_round_number_next_is_y(2, 3),
-            if_round_number_is_x_then_round_number_next_is_y(3, 4),
-            if_round_number_is_x_then_round_number_next_is_y(4, 5),
-            if_round_number_is_x_then_round_number_next_is_y(5, 6),
-            if_round_number_is_x_then_round_number_next_is_y(6, 7),
-            if_round_number_is_x_then_round_number_next_is_y(7, 8),
-            // if round number is 8, then round number next is 0 or 1
-            if_round_number_is_x_then_round_number_next_is_y(8, 0)
-                * (round_number_next - constant(1)),
-            // todo: The remaining 7·16 = 112 constraints are left as an exercise to the reader.
-        ]
+        // 2. if round number is 0, then next round number is 0
+        // DNF: rn in {1, ..., 9} \/ rn* = 0
+        let mut polynomial = (1..=NUM_ROUNDS + 1)
+            .map(|r| constant(r) - round_number.clone())
+            .fold(constant(1), MPolynomial::mul);
+        polynomial *= round_number_next.clone();
+        constraint_polynomials.push(polynomial);
+
+        // 3. if round number is 9, then next round number is 0 or 1
+        // DNF: rn =/= 9 \/ rn* = 0 \/ rn* = 1
+        polynomial = (0..=8)
+            .map(|r| constant(r) - round_number.clone())
+            .fold(constant(1), MPolynomial::mul);
+        polynomial *= constant(1) - round_number_next.clone();
+        polynomial *= round_number_next.clone();
+        constraint_polynomials.push(polynomial);
+
+        // 4. if round number is in {1, ..., 8} then next round number is +1
+        // DNF: (rn == 0 \/ rn == 9) \/ rn* = rn + 1
+        polynomial = round_number.clone()
+            * (constant(9) - round_number.clone())
+            * (round_number_next.clone() - round_number.clone() - constant(1));
+        constraint_polynomials.push(polynomial);
+
+        // Rescue-XLIX
+
+        // left-hand-side, starting at current round and going forward
+        let current_state: Vec<MPolynomial<XFieldElement>> = (0..STATE_SIZE)
+            .map(|i| variables[HashTableColumn::STATE0 as usize + i].clone())
+            .collect_vec();
+        let after_sbox = current_state.iter().map(|c| c.pow(ALPHA)).collect_vec();
+        let after_mds = (0..STATE_SIZE)
+            .map(|i| {
+                (0..STATE_SIZE)
+                    .map(|j| {
+                        MPolynomial::from_constant(
+                            BFieldElement::from(MDS[i * STATE_SIZE + j]).lift(),
+                            2 * FULL_WIDTH,
+                        ) * after_sbox[j].clone()
```

Comment:
Why not:
```suggestion
                    .map(|j| constant(MDS[i * STATE_SIZE + j] as usize) * after_sbox[j].clone())
```
(Both here and further below, too.)